test/hotspot/jtreg/runtime/handshake/HandshakeDirectTest.java

```diff
@@ -28,72 +28,52 @@
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller sun.hotspot.WhiteBox
+ * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI HandshakeDirectTest
+ * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -XX:GuaranteedSafepointInterval=10 -XX:+HandshakeALot -XX:+SafepointALot HandshakeDirectTest
  */
 
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.Semaphore;
+import sun.hotspot.WhiteBox;
 import java.io.*;
 
 public class HandshakeDirectTest  implements Runnable {
     static final int WORKING_THREADS = 32;
-    static final int DIRECT_HANDSHAKES_MARK = 500000;
+    static final int DIRECT_HANDSHAKES_MARK = 300000;
     static Thread[] workingThreads = new Thread[WORKING_THREADS];
-    static Semaphore[] handshakeSem = new Semaphore[WORKING_THREADS];
     static Object[] locks = new Object[WORKING_THREADS];
-    static boolean[] isBiased = new boolean[WORKING_THREADS];
     static AtomicInteger handshakeCount = new AtomicInteger(0);
 
     @Override
     public void run() {
         int me = Integer.parseInt(Thread.currentThread().getName());
+        WhiteBox wb = WhiteBox.getWhiteBox();
 
-        while (true) {
-            try {
-                if (!isBiased[me]) {
-                    handshakeSem[me].acquire();
-                    synchronized(locks[me]) {
-                        isBiased[me] = true;
-                    }
-                    handshakeSem[me].release();
-                }
-
+        while (handshakeCount.get() < DIRECT_HANDSHAKES_MARK) {
+            boolean walked = false;
+            synchronized(locks[me]) {
                 // Handshake directly some other worker
                 int handshakee = ThreadLocalRandom.current().nextInt(0, WORKING_THREADS - 1);
                 if (handshakee == me) {
                     // Pick another thread instead of me.
                     handshakee = handshakee != 0 ? handshakee - 1 : handshakee + 1;
                 }
-                handshakeSem[handshakee].acquire();
-                if (isBiased[handshakee]) {
-                    // Revoke biased lock
-                    synchronized(locks[handshakee]) {
-                        handshakeCount.incrementAndGet();
-                    }
-                    // Create new lock to be biased
-                    locks[handshakee] = new Object();
-                    isBiased[handshakee] = false;
-                }
-                handshakeSem[handshakee].release();
-                if (handshakeCount.get() >= DIRECT_HANDSHAKES_MARK) {
-                    break;
+                // Inflate locks[handshakee] if possible
+                System.identityHashCode(locks[handshakee]);
+                walked = wb.handshakeReadMonitors(workingThreads[handshakee]);
```

Comment:
It is not at all obvious that this revised test and the new WB routine actually test what was previously being tested. Do we actually need to involve monitors here or is that just something that has been picked to examine while in a handshake?

Reply:
So the purpose of this test was to exercise direct handshakes between threads. Back then I used biased locking because it was the only one that was using them (and I didn't thought about using whitebox). Since the test has proven to be good to uncover some bugs I tried to keep the same elements. Walking the stack looking for monitors is one. Then I added System.identityHashCode() to force transitions from stack-locked to inflated whenever possible so that we also change the state of the lock as we did with biased locking (from biased to stack-locked).